### PR TITLE
ENH: special: Add ufuncs ndtr_delta and trunc_ndtr; use trunc_ndtr for stats.truncnorm._cdf.

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
This is a proof-of-concept to resolve the performance issue reported in
https://github.com/scipy/scipy/issues/12733.

The essential code that implements stats.truncnorm._cdf is converted to two ufuncs,
`special.ndtr_delta` and `special.trunc_ndtr`, so the implementation of `truncnorm._cdf`
is now just

        def _cdf(self, x, a, b):
            return sc.trunc_ndtr(a, b, x)

A timing comparison: in master, with this setup:

    rng = np.random.default_rng(123456)
    a = rng.uniform(-10, 10, size=1000)
    b = a + rng.uniform(0, 10, size=1000)
    x = rng.normal(scale=10, size=1000)

I get

    In [54]: %timeit truncnorm.cdf(x, a, b)
    8.62 ms ± 66.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

With this PR, that becomes

    In [11]: %timeit truncnorm.cdf(x, a, b)
    191 µs ± 1.06 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

-----

Eventually it would be nice to replace the other methods in `truncnorm` that currently use slowish `nditer` loops with ufuncs.

### CODE REVIEW

**Comment by rgommers** [comment]
I'm not sure how common it is for us to add new functions into `scipy/special/cephes/` that aren't part of Cephes. Won't that be confusing? Unless we've done that regularly already, maybe best to either move the code or mark it clearly with code comments as not belonging to Cephes?

**Comment by WarrenWeckesser** [comment]
We've definitely edited the cephes files quite a bit, and I'd say we've been maintaining our own fork for many years.  I think of this code as `scipy-cephes`, far enough removed from the original that we should be able to continue editing and expanding the code as we see fit.  So I don't think there is any harm in adding new functions here.  Having said that, I'll take a look at moving the functions. I mainly put them there because it was convenient.

**Comment by rgommers** [comment]
> Having said that, I'll take a look at moving the functions.

No need in that case I'd say - I really was just asking without much context.

**Comment by person142** [comment]
A year or so ago I merged all C code into `cephes`; so it is now basically "the place to put C code for special functions".

File: scipy/special/cephes/ndtr.c [hunk 1]
```diff
@@ -224,6 +224,84 @@ double ndtr(double a)
...
+        sf_error("trunc_ndtr", SF_ERROR_DOMAIN, NULL);
+        return NPY_NAN;
+    }
+    if (x <= a) {
+        return 0.0;
+    }
+    if (x >= b) {
+        return 1.0;
+    }
+    delta = ndtr_delta(b, a);
```
Comment (mdhaber) [inline scipy/special/cephes/ndtr.c:283]: If I'm interpreting this correctly, the code below is for precision when there's subtractive cancellation, but it only runs when `delta` is exactly zero. What is delta is nearly zero but still has lost precision?

File: scipy/special/cephes/ndtr.c [hunk 2]
```diff
@@ -224,6 +224,84 @@ double ndtr(double a)
...
+
+    if (cephes_isnan(x) || cephes_isnan(a) || cephes_isnan(b) || (a >= b)) {
+        sf_error("trunc_ndtr", SF_ERROR_DOMAIN, NULL);
+        return NPY_NAN;
+    }
+    if (x <= a) {
+        return 0.0;
+    }
+    if (x >= b) {
+        return 1.0;
```
Comment (mdhaber) [inline scipy/special/cephes/ndtr.c:281]: If `b<a`, are these defined?

File: scipy/special/cephes/ndtr.c [hunk 3]
```diff
@@ -224,6 +224,84 @@ double ndtr(double a)
...
+    }
+    delta = ndtr_delta(b, a);
+    if (delta > 0) {
+        return ndtr_delta(x, a) / delta;
+    }
+    if (a < 0) {
```
Comment (mdhaber) [inline scipy/special/cephes/ndtr.c:287]: Originally I thought this code just ran when there was perfect subtractive cancellation, but it seems to run all the time given the order of arguments passed to `ndtr_delta`.

File: scipy/special/cephes/ndtr.c (hunk 1 above)
Comment (WarrenWeckesser) [reply scipy/special/cephes/ndtr.c:283]: Yup, that's a mistake.

I'm making some significant changes to the code that will make `ndtr_delta` much more accurate.  It's not a quick fix, though, so for now, I'm marking this PR as a draft.

**Comment by WarrenWeckesser** [comment]
Closed for renovations.

**Pull request closed without merging**